pandas/core/indexing.py

```diff
@@ -1746,7 +1746,12 @@ def _setitem_with_indexer_frame_value(self, indexer, value: "DataFrame"):
 
         unique_cols = value.columns.is_unique
 
-        if not unique_cols and value.columns.equals(self.obj.columns):
+        if name == "iloc":
+            for index, loc in zip(range(len(ilocs)), ilocs):
+                val = value.iloc[:, index]
+                self._setitem_single_column(loc, val, plane_indexer)
```

Comment:
can this be solved with an appropriate call to _align_frame or _align_series?  id really prefer to avoid passing `name` around.  by the time we get to setitem_with_indexer we should _always_ be working with positional indexers (i.e. iloc)

Reply:
Hm no not to my knowledge. The issue is, that we want to achieve oposing things between loc and iloc, but the objects do look the same. ``setitem_with_indexer`` could be solved with a default value, that is not a problem. 

But I can not see a way to do this without passing some kind of flag to the align calls.

Reply:
Looked into this a bit again. The indexer in ``setitem_with_indexer`` already is positional, but the value may not be properly aligned for loc, hence we call the align functions there. This is not necessary for the iloc path, but this information is lost there, if we do not pass it through somehow.

Reply:
it sounds like we should be doing some appropriate alignment within Loc before passing it into iLoc?

Reply:
That would be an option, but would result in a bigger rewrite and I am not sure if that would introduce more complexity, because the ``align_series`` method is called in various places. Should I give it a try?

Reply:
it looks like this PR is making it so that we _never_ call align_frame/align_series if setitem_with_indexer was called from iloc, only if it was called from loc.  setitem_with_indexer is only called from one place within loc.  Can we just do that alignment right before that call?

Reply:
Hm, we could try, but it is not that simple. For example before calling `` setitem_with_indexer `` from loc, ``value`` may be a ``DataFrame`` with different dtypes. Currently ``_align_series`` is called for every column to get a ``ndarray`` with the correct dtype. If we call this before calling ``setitem_with_indexer`` we would have to call ``_align_frame`` or copy the logic, which decides if we align every colum at once or separately. ``_align_frame`` returns an object dtype array, which obviously would introduce a lot of dtype errors. That is what I meant with a bigger rewrite here.

Reply:
thanks for explaining, i think i have a better handle on it now.

ive got an upcoming PR to make all 2D cases go through split_path.  let's revisit this after that goes through and see if we can make it cleaner

Reply:
That sounds great. Just ping me when merged, then I will have a look.